### PR TITLE
Echo parsing

### PR DESCRIPTION
**Pull request by wasade**

### CODE REVIEW

**Comment by wasade** [comment]
Thanks :)

File: knimin/lib/parse.py [hunk 1]
```diff
@@ -94,3 +95,84 @@ def parse_qpcr_object(contents):
...
+    int
+        The index of the first empty line following the section, or the index
+        position of the end of the file.
+
+    Raises
+    ------
+    ValueError
+        If the section is not found
+    """
+    start = -1
```
Comment (qiyunzhu) [inline knimin/lib/parse.py]: I feel that this calculation is overly complicated. Maybe possible to interate the `lines` only once. For example:
```
start = -1
end = len(lines)
started = False
for idx, line in enumerate(lines):
    if line.startswith(label):
        start = idx
        started = True
    elif started and not line:
        end = idx
        break
if start == -1:
    raise ValueError('%s section appears to be missing' % label)
```

File: knimin/lib/parse.py [hunk 2]
```diff
@@ -94,3 +95,84 @@ def parse_qpcr_object(contents):
...
+
+    Raises
+    ------
+    ValueError
+        If the header is missing
+        If the exceptions section is missing
+        If the details section is missing
+    """
+    data = data.splitlines()
+    if not data[0].startswith('Run ID'):
```
Comment (qiyunzhu) [inline knimin/lib/parse.py]: Is there a real chance that the `data` is not formatted correctly?

File: knimin/lib/tests/test_parse.py [hunk 3]
```diff
@@ -97,6 +101,115 @@ def test_parse_qpcr_object(self):
             "u'Status']")
 
 
+class EchoParserTests(TestCase):
+    def setUp(self):
+        base = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'data')
```
Comment (qiyunzhu) [inline knimin/lib/tests/test_parse.py]: I wonder if it's easier to do `from os.path import join, dirname, realpath` then do `join` directly, like other scripts in the same project did.

File: knimin/lib/parse.py [hunk 4]
```diff
@@ -94,3 +95,84 @@ def parse_qpcr_object(contents):
...
+        A DataFrame composed of the [DETAILS] section
+
+    Raises
+    ------
+    ValueError
+        If the header is missing
+        If the exceptions section is missing
+        If the details section is missing
+    """
+    data = data.splitlines()
```
Comment (qiyunzhu) [inline knimin/lib/parse.py]: If `data` has extra trailing empty new lines, this conversion will leave some empty elements in the end. Maybe  do `data.rstrip('\n')` beforehand?

File: knimin/lib/parse.py (hunk 1 above)
Comment (wasade) [inline knimin/lib/parse.py]: I'd argue that is more complex...

File: knimin/lib/parse.py (hunk 4 above)
Comment (wasade) [inline knimin/lib/parse.py]: that shouldn't be an issue

File: knimin/lib/parse.py (hunk 2 above)
Comment (wasade) [inline knimin/lib/parse.py]: I don't know

File: knimin/lib/tests/test_parse.py (hunk 3 above)
Comment (wasade) [inline knimin/lib/tests/test_parse.py]: no strong feelings

**Comment by wasade** [comment]
Failures look unrelated

File: knimin/lib/parse.py (hunk 1 above)
Comment (qiyunzhu) [inline knimin/lib/parse.py]: Why? That saves one loop thru the lines, which is expensive.

Comment (wasade) [inline knimin/lib/parse.py]: This is not performance bound code, and the total number of iterations is the same as at most, the code will the entire file only once as the forloops do not overlap. The multiple loops also represent a logical separation, and I'd argue is more in line with KISS, where each forloop statement does one specific thing.

Comment (qiyunzhu) [inline knimin/lib/parse.py]: Oh I see. Thanks for explanation!

**Pull request merged**